src/jit/optimizer.cpp

```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
                 }
 #endif // FEATURE_EH_FUNCLETS && defined(_TARGET_ARM_)
 
-                /* At this point we have a loop - record it in the loop table
+                // Compact the loop (sweep through it and move out any blocks that aren't part of the
+                // flow cycle), and count the exits.
+                BasicBlock* moveAfter = nullptr;
+                for (BasicBlock* previous = top->bbPrev; previous != bottom;)
+                {
+                    BasicBlock* block = previous->bbNext;
+
+                    if (!loopBlocks.isMember(block->bbNum))
+                    {
+                        // This code is lexically between TOP and BOTTOM, but it does not
+                        // participate in the flow cycle.  Check for a run of consecutive
+                        // such blocks.
+                        BasicBlock* lastExitBlock = block;
+                        BasicBlock* nextLoopBlock = block->bbNext;
+                        while (!loopBlocks.isMember(nextLoopBlock->bbNum))
+                        {
+                            lastExitBlock = nextLoopBlock;
+                            nextLoopBlock = nextLoopBlock->bbNext;
+                            // This loop must terminate because we know BOTTOM is in loopBlocks.
+                        }
+
+                        if (moveAfter == nullptr)
+                        {
+                            // Find an insertion point for blocks we're going to move.  Move them down
+                            // out of the loop, and if possible find a spot that won't break up fall-through.
+                            moveAfter = bottom;
+                            while (moveAfter->bbFallsThrough())
+                            {
+                                // Avoid splitting up this fall-through if we can.
+                                BasicBlock* next = moveAfter->bbNext;
+
+                                if (!BasicBlock::sameEHRegion(moveAfter, next))
+                                {
+                                    // Don't cross an EH region boundary
+                                    break;
+                                }
+
+                                if ((next->bbJumpKind == BBJ_ALWAYS) || (next->bbJumpKind == BBJ_COND))
+                                {
+                                    unsigned int destNum = next->bbJumpDest->bbNum;
+                                    if ((destNum >= top->bbNum) && (destNum <= bottom->bbNum) &&
+                                        !loopBlocks.isMember(destNum))
+                                    {
+                                        // Reversing this branch out of block `next` could confuse this algorithm, so
```

Comment:
> Those branches would still be reversed if we moved the blocks anywhere after next.

That's the bit I'm missing. If the code can find a fallthrough, then the "fixup" for the in loop blocks is something it can already handle, so I don't see yet how skipping past a problematic next to some non-fallthrough spot below it wouldn't work. Maybe you just need to be a bit more explicit on what exactly about reversal leads to confusion.

> Regarding "if you find one, use it, otherwise just insert after BOTTOM", were you just sketching out how an algorithm that kept walking would work, or do you have a reason to think that BOTTOM is a better insertion point than the last legal block?

Just thinking how I would have started writing this bit. I probably would have tried to leverage `fgFindInsertPoint` to do the actual searching, and looking at it now, it seems like there might be logic there you need here, eg a BBJ_CALLFINALLY does not fall through but you can't safely move new blocks after it.

As far as a "best" insertion point -- not creating a new block is preferable from a TP standpoint. Other than that I would say try and avoid creating new lexically backwards branches. Insert after BOTTOM would guarantee this, though you could get the same guarantee if you moved the run of blocks before any post-loop block they reach, and if there is just one such block and maybe try and move the run just before it.

I suppose it would also be nice to avoid moving the blocks into another loop where you'll probably just end up moving them again later on, but that might be tricky to pull off or maybe doesn't happen with any frequency. You could perhaps use `BBF_BACKWARD_JUMP` as a screen.

None of this is all that important, if what you have now is correctly moving blocks. If so I would not worry as much about trying to make it optimal, unless you already see issues that come from particular placement choices.